File: firebase/functions/lib/events/create_event.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'live_meetings/breakouts/check_hostless_go_to_breakouts.dart';
import '../on_call_function.dart';
import 'notifications/event_emails.dart';
import '../utils/emulator_utils.dart';
import '../utils/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/community/membership.dart';

/// This function handles events after event creation
class CreateEvent extends OnCallMethod<CreateEventRequest> {
  EventEmails eventEmailUtils;
  CheckHostlessGoToBreakouts checkHostlessGoToBreakouts;

  CreateEvent({
    EventEmails? eventEmailUtils,
    CheckHostlessGoToBreakouts? checkHostlessGoToBreakouts,
  })  : eventEmailUtils = eventEmailUtils ?? EventEmails(),
        checkHostlessGoToBreakouts =
            checkHostlessGoToBreakouts ?? CheckHostlessGoToBreakouts(),
        super(
          CreateEventRequest.functionName,
          (jsonMap) => CreateEventRequest.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    CreateEventRequest request,
    CallableContext context,
  ) async {
    final event = await firestoreUtils.getFirestoreObject(
      path: request.eventPath,
      constructor: (map) => Event.fromJson(map),
    );

    final membershipDoc = await firestore
        .document(
          'memberships/${context.authUid}/community-membership/${event.communityId}',
        )
        .get();
    final membership = Membership.fromJson(
      firestoreUtils.fromFirestoreJson(membershipDoc.data.toMap() ?? {}),
    );
    final isModOrCreator =
        event.creatorId == context.authUid || membership.isMod;

    if (!isEmulator && !isModOrCreator) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    await _handleEmailNotifications(event);

    if (event.eventType == EventType.hostless) {
      await checkHostlessGoToBreakouts.enqueueScheduledCheck(event);
    }
  }

  Future<void> _handleEmailNotifications(Event event) async {
    // Send initial sign up email to user.
    await eventEmailUtils.sendEmailsToUsers(
      eventPath: event.fullPath,
      userIds: [event.creatorId],
      emailType: EventEmailType.initialSignUp,
    );

    await eventEmailUtils.enqueueReminders(event);
  }
}

File: firebase/functions/src/events/create_event.ts
import { OnCallMethod } from '../on_call_function'
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils'
import { isEmulator } from '../utils/infra/emulator_utils'
import {
    Event,
    Membership,
    EventType,
    getEventType,
    getEventFullPath,
    EventEmailType,
} from '../types'
import { membershipIsMod } from '../types'
import { EventEmails } from './notifications/event_emails'
import { CheckHostlessGoToBreakouts } from './live_meetings/breakouts/check_hostless_go_to_breakouts'
import * as functions from 'firebase-functions'

interface CreateEventRequest {
    eventPath: string
}

export class CreateEvent extends OnCallMethod<CreateEventRequest> {
    private eventEmailUtils: EventEmails
    private checkHostlessGoToBreakouts: CheckHostlessGoToBreakouts

    constructor() {
        super('CreateEvent', (jsonMap) => jsonMap as CreateEventRequest)
        this.eventEmailUtils = new EventEmails()
        this.checkHostlessGoToBreakouts = new CheckHostlessGoToBreakouts()
    }

    async action(
        request: CreateEventRequest,
        context: functions.https.CallableContext
    ): Promise<void> {
        const event: Event = await firestoreUtils.getFirestoreObject({
            path: request.eventPath,
            constructor: (map) => map as unknown as Event,
        })

        const membershipDoc = await firestore
            .doc(`memberships/${context.auth?.uid}/community-membership/${event.communityId}`)
            .get()
        const membership: Membership = firestoreUtils.fromFirestoreJson(
            membershipDoc.data() ?? {}
        ) as unknown as Membership
        const isModOrCreator =
            event.creatorId === context.auth?.uid || membershipIsMod(membership.status)

        if (!isEmulator && !isModOrCreator) {
            throw new functions.https.HttpsError('failed-precondition', 'unauthorized')
        }

        await this._handleEmailNotifications(event)

        if (getEventType(event) === EventType.hostless) {
            await this.checkHostlessGoToBreakouts.enqueueScheduledCheck(event)
        }
    }

    private async _handleEmailNotifications(event: Event): Promise<void> {
        await this.eventEmailUtils.sendEmailsToUsers({
            eventPath: getEventFullPath(event),
            userIds: [event.creatorId!],
            emailType: EventEmailType.initialSignUp,
        })

        await this.eventEmailUtils.enqueueReminders(event)
    }
}
